probe_scraper/scrapers/buildhub.py

```diff
@@ -0,0 +1,159 @@
+from datetime import datetime
+import pprint
+import requests
+import re
+
+
+class NoDataFoundException(Exception):
+    pass
+
+
+class Buildhub(object):
+
+    search_url = "https://buildhub.moz.tools/api/search"
+    default_window = 1000
+
+    date_formats = ("%Y-%m-%dT%H:%M:%SZ", "%Y-%m-%dT%H:%M:%S.%f")
+
+    def _paginate_revision_dates(self, iteration, channel, min_version, product,
+                                 locale, platform, max_version, verbose, window):
+        query_str = [
+            {"term": {"source.product": product}},
+            {"term": {"target.channel": channel}},
+            {"range": {"target.version": {"gte": str(min_version)}}},
+            {"term": {"target.locale": locale}},
+            {"term": {"target.platform": platform}}
+        ]
+
+        if max_version is not None:
+            query_str.append({
+                "bool": {
+                    "should": [
+                        {"range": {"target.version": {"lte": str(max_version)}}},
+                        {"prefix": {"target.version": str(max_version)}}
+                    ]
+                }
+            })
+
+        body = {
+            "query": {
+                "bool": {
+                    "filter": query_str
+                }
+            },
+            "size": window
+        }
+
+        if iteration != 0:
+            body["from"] = iteration * window
+
+        if verbose:
+            print("------QUERY STRING------\n")
+            pprint.pprint(body)
+
+        response = requests.post(url=Buildhub.search_url, json=body)
+        data = response.json()
+
+        if verbose:
+            print("------QUERY RESULTS------\n")
+            pprint.pprint(data)
+
+        return data
+
+    def _distinct_and_clean(self, records):
+        """
+        For more information on the schema of the records,
+        see the Buildhub API documentation:
+        https://buildhub.readthedocs.io/en/latest/api.html#more-about-the-data-schema
+        """
+        cleaned_records = {}
+
+        for record in records:
+            # %:z not supported, see https://bugs.python.org/msg169952
+            # Drop the tz portion entirely
+            d = record["_source"]["download"]["date"]
+            if re.search(r"\+\d{2}:\d{2}$", d):
+                d = d[:-6]
+
+            date = None
+            try:
+                date = datetime.strptime(d, Buildhub.date_formats[0])
+            except ValueError:
+                pass
+
+            if date is None:
+                date = datetime.strptime(d, Buildhub.date_formats[1])
```

Comment:
Maybe put this parsing in its own function so you could just return from a for loop on the first format that works?

Also, you should probably use `self.date_formats` here -- no need to name the class by name.

Reply:
My only hesitancy with using a loop was I want the last element to fail if it doesn't parse. That will indicate to future us that they changed the format of the date string, and with only two elements it seems simple enough to unroll?